File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
from collections import defaultdict

def solution(genres, plays):
    answer = []
    
    # 각 장르별 재생시간 저장하기 위해 defaultdict 사용
    dic = defaultdict(int)
    
    # 장르, 재생시간, 고유번호 저장 배열
    array = []
    
    for i in range(len(genres)):
        # 각 장르별 재생시간 총합
        dic[genres[i]] += plays[i]
        # [장르, 재생시간, 고유번호 저장]
        array.append([genres[i], plays[i], i])
        
    # 장르 재생시간을 기준으로 내림차순 정렬
    sdic = sorted(dic.items(), key = lambda x: x[1], reverse=True)
    
    # [장르, 재생시간, 고유번호] 정렬
    array = sorted(array, key = lambda x : (x[0], -x[1], x[2]))
    
    for i in range(len(sdic)):
        # 최대 두 개씩만 앨범에 들어갈 수 있으니 카운트
        count = 2
        for j in range(len(array)):
            if sdic[i][0] == array[j][0] and count != 0:
                answer.append(array[j][2])
                count -= 1
                
    return answer

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;
import java.io.*;

class Solution {
    static int[] answer;
    static ArrayList<String> gList;
    static HashMap<String, ArrayList<Info>> hm;
    static HashMap<String, Long> total;
    
    static class Info {
        int count, index;
        
        Info(int c, int i) {
            count = c;
            index = i;
        }
    }
    
    public int[] solution(String[] genres, int[] plays) {
        hm = new HashMap<>();
        total = new HashMap<>();
        gList = new ArrayList<>();
        
        for (int i = 0; i < genres.length; i++) {
            ArrayList<Info> list = new ArrayList<>();
            String g = genres[i];
            
            if (hm.containsKey(g)) list = hm.get(g);
            list.add(new Info(plays[i], i));
            hm.put(g, list);
        }
        
        int musicCnt = 0;
        for (String g : hm.keySet()) {
            long cnt = 0;
            for (Info info : hm.get(g)) cnt += info.count;
            total.put(g, cnt);
            gList.add(g);
            
            ArrayList<Info> infos = hm.get(g);
            infos.sort((i1, i2) -> {
                if (i1.count == i2.count) {
                    return i1.index - i2.index;
                }
                return i2.count - i1.count;
            });
            
            int infoSize = infos.size();
            if (infoSize >= 2) infoSize = 2;
            musicCnt += infoSize;
        }
        
        gList.sort((g1, g2) -> Long.compare(total.get(g2), total.get(g1)));
        
        answer = new int[musicCnt];
        int Index = 0;
        for (String g : gList) {
            for (int i = 0; i < Math.min(hm.get(g).size(), 2); i++) {
                // System.out.println("index : " + Index++);
                answer[Index++] = hm.get(g).get(i).index;
            }
        }
        
        return answer;
    }
}